# HAMOYE ONESCHOOL
## TRACK: DATA SCIENCE
## STAGE B:---Regression in Machine Learning Quiz(12-20)¶
# BY: KATHERINE DAVID
# ID: 15722cdf5c81f000

In [1]:
#import libraies
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [40]:
#load dataset
df=pd.read_csv('energydata_complete.csv')

In [3]:
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [41]:
#drop the date column
df.drop('date',axis=1,inplace=True)

# Question 12
From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [5]:
from sklearn.preprocessing import MinMaxScaler
scaler= MinMaxScaler()
normalized_df= pd.DataFrame(scaler.fit_transform(df),columns=df.columns)

In [6]:
x=normalized_df[['T2']]
y=normalized_df[['T6']]

In [7]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test= train_test_split(x,y,test_size=0.3,random_state=42)


In [8]:
from sklearn.linear_model import LinearRegression
ml=LinearRegression()

ml.fit(x_train,y_train)


LinearRegression()

In [9]:
prediction=ml.predict(x_test)

In [10]:
from sklearn.metrics import r2_score
r2= r2_score(y_test,prediction)
round(r2,2)

0.64

ANSWER:- The R^2 value is 0.64

###### Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. Answer the following questions:

In [11]:
normalized_df.drop('lights',axis=1,inplace=True)

In [12]:
new_x=normalized_df.drop('Appliances',axis=1)
new_y=normalized_df['Appliances']

In [13]:
x_train,x_test,y_train,y_test= train_test_split(new_x,new_y,test_size=0.3,random_state=42)


In [14]:
ml_pred=LinearRegression()
ml_pred.fit(x_train,y_train)

LinearRegression()

In [15]:
pred_test=ml_pred.predict(x_test)

# Question 13
What is the Mean Absolute Error (in two decimal places)?

In [16]:
from sklearn.metrics import mean_absolute_error
mae=mean_absolute_error(y_test,pred_test)
round(mae,2)

0.05

ANSWER:- The Mean Absolute Error is 0.05

# Question 14
What is the Residual Sum of Squares (in two decimal places)?

In [17]:
rss=np.sum(np.square(y_test-pred_test))
round(rss,2)

45.35

ANSWER:- The Residual Sum of Squares is 45.35

# Question 15

What is the Root Mean Squared Error (in three decimal places)?

In [18]:
from sklearn.metrics import mean_squared_error
rmse= np.sqrt(mean_squared_error(y_test,pred_test))
round(rmse,3)

0.088

ANSWER:- The Root Mean Squared Error is 0.088

# Question 16
What is the Coefficient of Determination (in two decimal places)?

In [19]:
from sklearn.metrics import r2_score
r2= r2_score(y_test,pred_test)
round(r2,2)

0.15

ANSWER:- The Coefficient of Determination is 0.15

# Question 17
Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [32]:
def get_weights_df(ml_pred,feat,col_name):
    weights= pd.Series(ml_pred.coef_,feat.columns).sort_values()
    weights_df=pd.DataFrame(weights).reset_index()
    weights_df.columns=['Features',col_name]
    weights_df[col_name].round(3)
    return weights_df
linear_model_weights = get_weights_df(ml_pred, x_train, 'Linear_Model_Weight')
linear_model_weights

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv2             0.000770
11          rv1             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

ANSWER:- The features with the lowest and highest weights are RH_2,RH_1

# Question 18
Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [33]:
from sklearn.linear_model import Ridge
reg=Ridge(alpha=0.4)

In [34]:
reg.fit(x_train,y_train)

Ridge(alpha=0.4)

In [35]:
reg_pred=reg.predict(x_test)
reg_rmse= np.sqrt(mean_squared_error(y_test,reg_pred))
round(reg_rmse,3)

0.088

ANSWER:- No there was no change to the root mean squared error (RMSE) when evaluated on the test set 

# Question 19
Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [38]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

def get_weights_df(ml_pred, feat, col_name):
    

    weights = pd.Series(ml_pred.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')
non_zero_weights=lasso_weights_df[lasso_weights_df['Lasso_weight']!=0]
non_zero_weights

Features  Lasso_weight
0      RH_out     -0.049557
1        RH_8     -0.000110
24  Windspeed      0.002912
25       RH_1      0.017880

ANSWER:- 4 features have non-zero feature weights

# Question 20
What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)

In [39]:
lasso_pred=lasso_reg.predict(x_test)
lasso_rmse=np.sqrt(mean_squared_error(y_test,lasso_pred))
round(lasso_rmse,3)

0.094

ANSWER:- The new RMSE with lasso regression is 0.094